### PR TITLE
Add Support for migrating Schema/Catalog ACL for Interactive cluster

### LINKED ISSUE #1192: [FEATURE]: Create External Location ACL for Principals based on Interactive cluster permission using SPN
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Currently Users access interactive clusters using SPN configured which has permission on the underlying azure storage.
When migrating to UC, we need to create permission to interactive clusters users on the external location, catalog and schema  based on the underlying spn access on the storage

Related issues:

- https://github.com/databrickslabs/ucx/issues/907

### Proposed Solution

 - identify interactive clusters with spn configured.
 - check spn permission on the underlying storage
 - grant appropriate permission on the external location to the users of the interactive clusters
 - grant use permission on the catalog and schema equivalent tin UC for the users of the interactive clusters

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by HariGS-DB**
## Changes
 - Identified all the database acl grants from the PrincipalACL class and filters for only database grants
 - creates the mapping from hive metastore schema to UC schema and catalog using the Table Mapping
 - replaces the hive metastore action with equivalent UC action for both schema and catalog
 - handles both aws and azure
 - Note: This PR doesn't cover external location permission which is part of the same issue. I will submit separate PR for that

### Linked issues

Fixes Partially #1192  and #1193

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [X] manually tested
- [X] added unit tests
- [X] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 1]
```diff
@@ -23,6 +32,30 @@ def create_all_catalogs_schemas(self, prompts: Prompts):
...
                 self._create_schema(candidate_catalog, candidate_schema)
+        self._update_principal_acl()
+
+
+    def _get_catalog_schema_grants(self):
+        catalog_grants: set[Grant] = set()
+        src_trg_schema_mapping = self._get_src_database_target_mapping()
+        grants = self._principal_grants.get_interactive_cluster_grants()
+        database_grants = [grant for grant in grants if grant.table is None or grant.view is None]
+        new_grants = [dataclasses.replace(db_grant, database=src_trg_schema_mapping[db_grant.database]['target_schema'],catalog=src_trg_schema_mapping[db_grant.database]['target_catalog']) for db_grant in database_grants]
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:43]: If list comprehension doesn't fit 32-48 characters, rewrite it as a for loop to get more readability

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 2]
```diff
@@ -23,6 +32,30 @@ def create_all_catalogs_schemas(self, prompts: Prompts):
...
             for candidate_schema in schemas:
                 self._create_schema(candidate_catalog, candidate_schema)
+        self._update_principal_acl()
+
+
+    def _get_catalog_schema_grants(self):
+        catalog_grants: set[Grant] = set()
+        src_trg_schema_mapping = self._get_src_database_target_mapping()
+        grants = self._principal_grants.get_interactive_cluster_grants()
+        database_grants = [grant for grant in grants if grant.table is None or grant.view is None]
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:42]: There's type_and_key() method on a grant that you should be using

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 3]
```diff
@@ -23,6 +32,30 @@ def create_all_catalogs_schemas(self, prompts: Prompts):
...
+        new_grants = [dataclasses.replace(db_grant, database=src_trg_schema_mapping[db_grant.database]['target_schema'],catalog=src_trg_schema_mapping[db_grant.database]['target_catalog']) for db_grant in database_grants]
+        for grant in new_grants:
+            catalog_grants.add(dataclasses.replace(grant, database=""))
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:45]:
```suggestion
            catalog_grants.add(dataclasses.replace(grant, database=None))
```

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 4]
```diff
@@ -71,6 +104,16 @@ def _list_target(self) -> tuple[set[str], dict[str, set[str]]]:
             target_schemas[target_catalog].add(target_schema)
         return target_catalogs, target_schemas
 
+    def _get_src_database_target_mapping(self) -> dict[str, dict]:
+        """generate a list of catalogs and schema to be created from table mappings."""
+        src_trg_schema_mapping: dict[str, dict] = {}
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:109]: What is trg? dst?

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 5]
```diff
@@ -71,6 +104,16 @@ def _list_target(self) -> tuple[set[str], dict[str, set[str]]]:
             target_schemas[target_catalog].add(target_schema)
         return target_catalogs, target_schemas
 
+    def _get_src_database_target_mapping(self) -> dict[str, dict]:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:107]: Why this method is defined so far away from _get_catalog_schema_grants?

**Review by nfx — CHANGES_REQUESTED** [review]
Few nits. And need tests

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py (hunk 4 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/catalog_schema.py]: its the mapping from src hivemetastore schema to its target uc catalog and UC schema. have added a comment to make it clear

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py (hunk 2 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/catalog_schema.py]: I taking the output from self._principal_grants.get_interactive_cluster_grants() and filtering for only database level grants

**Review by nfx — CHANGES_REQUESTED** [review]

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**